File: apps/android/app/src/main/kotlin/xyz/aethelred/wallet/core/audit/AuditEvent.kt
package xyz.aethelred.wallet.core.audit

import kotlinx.serialization.Serializable

/**
 * Discriminator for [AuditEvent]. Mirrors the TypeScript union in
 * `packages/audit/src/types.ts` — any new value here must have a
 * corresponding string literal on the TS side so the control-plane
 * ingestion pipeline decodes it without adaptation.
 */
@Serializable
public enum class AuditEventKind(public val wire: String) {
    REQUEST_RECEIVED("request-received"),
    POLICY_EVALUATED("policy-evaluated"),
    APPROVAL_REQUESTED("approval-requested"),
    APPROVAL_DECIDED("approval-decided"),
    SIGNING_EXECUTED("signing-executed"),
    RESPONSE_SENT("response-sent"),
    SESSION_CREATED("session-created"),
    SESSION_REVOKED("session-revoked"),
    WORKSPACE_SWITCHED("workspace-switched"),
    ACCOUNT_CREATED("account-created"),
    ACCOUNT_IMPORTED("account-imported"),
    KEY_GENERATED("key-generated"),
    LOCK_STATE_CHANGED("lock-state-changed"),
    WALLET_INITIALIZED("wallet-initialized"),
    EXPORT_REQUESTED("export-requested"),
    CREDENTIAL_ENROLLED("credential-enrolled"),
    CREDENTIAL_VERIFIED("credential-verified"),
    CREDENTIAL_VERIFICATION_FAILED("credential-verification-failed"),
    CREDENTIAL_REVOKED("credential-revoked");

    public companion object {
        /** Reverse lookup for decoding wire kinds into the enum. */
        public fun fromWire(value: String): AuditEventKind? =
            entries.firstOrNull { it.wire == value }
    }
}

/**
 * Hash-linked audit event identical in shape to the TypeScript
 * `AuditEvent` interface. JSON-serializable via kotlinx.serialization.
 *
 * @property id Unique event identifier (`evt-<uuid>`).
 * @property sequenceNumber Monotonic counter inside a single subject/workspace.
 * @property timestamp Unix millisecond timestamp of creation.
 * @property kind Audit kind discriminator.
 * @property subjectId Subject (person/service/agent) that performed the action.
 * @property workspaceId Workspace the action ran inside.
 * @property appId Optional connected-app identifier.
 * @property sessionId Optional connect-session identifier.
 * @property intentId Optional signing-intent identifier.
 * @property detail Free-form per-kind payload.
 * @property previousHash Hex SHA-256 of the preceding event in the chain.
 * @property eventHash Hex SHA-256 of this event's canonicalized bytes.
 */
@Serializable
public data class AuditEvent(
    public val id: String,
    public val sequenceNumber: Long,
    public val timestamp: Long,
    public val kind: AuditEventKind,
    public val subjectId: String,
    public val workspaceId: String,
    public val appId: String? = null,
    public val sessionId: String? = null,
    public val intentId: String? = null,
    public val detail: Map<String, String> = emptyMap(),
    public val previousHash: String,
    public val eventHash: String,
)

File: apps/ios/AethelredWallet/Core/Audit/AuditEvent.swift
import Foundation

/// Audit-event kinds — shape-for-shape mirror of
/// `packages/audit/src/types.ts::AuditEventKind`.
///
/// The string values must match the TypeScript definitions exactly so
/// events produced by the iOS app can be replayed through the Elixir
/// event-ingestion pipeline without translation.
public enum AuditEventKind: String, Codable, Sendable, Equatable, CaseIterable {
    case requestReceived = "request-received"
    case policyEvaluated = "policy-evaluated"
    case approvalRequested = "approval-requested"
    case approvalDecided = "approval-decided"
    case signingExecuted = "signing-executed"
    case responseSent = "response-sent"
    case sessionCreated = "session-created"
    case sessionRevoked = "session-revoked"
    case workspaceSwitched = "workspace-switched"
    case accountCreated = "account-created"
    case accountImported = "account-imported"
    case keyGenerated = "key-generated"
    case lockStateChanged = "lock-state-changed"
    case walletInitialized = "wallet-initialized"
    case exportRequested = "export-requested"
    case credentialEnrolled = "credential-enrolled"
    case credentialVerified = "credential-verified"
    case credentialVerificationFailed = "credential-verification-failed"
    case credentialRevoked = "credential-revoked"
}

/// Mirror of `AuditEvent` in the TypeScript audit package. Fields use
/// the same names and ordering so JSON encoding is identical — tests in
/// `AuditCaptureTests.swift` assert round-trip compatibility against the
/// TS canonical envelope.
public struct AuditEvent: Codable, Sendable, Identifiable, Equatable {
    public let id: String
    public let sequenceNumber: Int
    public let timestamp: Int64
    public let kind: AuditEventKind
    public let subjectId: String
    public let workspaceId: String
    public let appId: String?
    public let sessionId: String?
    public let intentId: String?
    public let detail: [String: String]
    public let previousHash: String
    public let eventHash: String

    public init(
        id: String,
        sequenceNumber: Int,
        timestamp: Int64,
        kind: AuditEventKind,
        subjectId: String,
        workspaceId: String,
        appId: String? = nil,
        sessionId: String? = nil,
        intentId: String? = nil,
        detail: [String: String] = [:],
        previousHash: String,
        eventHash: String
    ) {
        self.id = id
        self.sequenceNumber = sequenceNumber
        self.timestamp = timestamp
        self.kind = kind
        self.subjectId = subjectId
        self.workspaceId = workspaceId
        self.appId = appId
        self.sessionId = sessionId
        self.intentId = intentId
        self.detail = detail
        self.previousHash = previousHash
        self.eventHash = eventHash
    }
}

/// Evidence record — bundle of audit events exported for compliance.
public struct EvidenceRecord: Codable, Sendable, Equatable {
    public let id: String
    public let title: String
    public let events: [AuditEvent]
    public let chainValid: Bool
    public let createdAt: Int64
    public let exportedAt: Int64?
}

/// Export payload shape expected by the control plane.
public struct ExportPackage: Codable, Sendable, Equatable {
    public let version: String
    public let exportedAt: Int64
    public let events: [AuditEvent]
    public let integrityHash: String
    public let chainStartSequence: Int
    public let chainEndSequence: Int
}
